lib/pool/ConnectionPool.js

```diff
@@ -60,7 +60,17 @@ class ConnectionPool extends BaseConnectionPool {
     const { id } = connection
     debug(`Marking as 'dead' connection '${id}'`)
     if (this.dead.indexOf(id) === -1) {
-      this.dead.push(id)
+      // It might happen that `markDead` is called jsut after
+      // a pool update, and in such case we will add to the dead
+      // list a node that no longer exist. The following check verify
+      // that the connection is still part of the pool before
+      // marking it as dead.
+      for (var i = 0; i < this.size; i++) {
+        if (this.connections[i].id === id) {
+          this.dead.push(id)
+          break
+        }
+      }
```

Comment:
If not added to `dead`, it still continues on to set `.status`, `.deadCount`, `.resurrectTimeout`, or call `.sort()`.
Would it be better to not do these and return `this` right after?

Reply:
That's true, but it's not a problem as the connection will be garbage collected at the end of the function :)
There is no need to add an "exists" logic just to add an early return imho.